### PR TITLE
Expose default database collation through 'sql-database-projects' extension API

### PR DESCRIPTION
**Pull request by nahk-ivanov**
For the purpose of schema conversion we would need to know whether target database is configured as CI or CS. This will be used to produce a warning, if we detect a case-sensitive identifier, but database is configured as CI. In order to support this scenario we need to access `<DefaultCollation/>` property of the project.

This change adds new method to the `ISqlProject` interface that allows to read the value of the project property. There already was similar method for the SQL version/platform (`<DSP/>` property) and while working on the change, I took an opportunity to refactor the way project properties are extracted from the XML. Original code was hardcoded in the `getProjectTargetVersion` and I extracted it into separate `evaluateProjectPropertyValue` helper, that can be used in the future by any getter or access method that is supposed to return a value of the single property. This also allows us to improve the way properties are retrieved from the XML. Today the logic is very rudimentary - we read the first matching XML element with the required name. This is not correct as it does not verify the parent to be `<PropertyGroup/>`, neither it evaluates the `Condition` attributes nor property value itself. I did not invest in this, but added a TODO there explaining that the method may not perform as expected.

After the helper method was added, I updated the existing `getProjectTargetVersion` method to leverage it. The only complication here was the error throwing logic, as it was using custom error message. I preserved that, as there were tests verifying it already. For the new accessor method I did not introduce a special error message and rely on generic one I defined for use within the helper method. Additionally, for the collation we return default value of `SQL_Latin1_General_CP1_CI_AS`, if project does not have the property defined. This is what SSDT for Visual Studio shows in the UI when property is missing and I decided to align with that.

Finally, I added tests for both - original `getProjectTargetVersion` and new collation extraction method to make sure they work as expected. While working on the tests, I've noticed complaints that some rehected promises were not awaited. I added missing `await`s and explained the scenario there in the comment.

### CODE REVIEW

File: extensions/sql-database-projects/src/common/constants.ts [hunk 1]
```diff
@@ -203,6 +203,7 @@ export function folderAlreadyExists(filename: string) { return localize('folderA
 export function fileAlreadyAddedToProject(filepath: string) { return localize('fileAlreadyAddedToProject', "A file with the path '{0}' has already been added to the project", filepath); }
 export function folderAlreadyAddedToProject(folderpath: string) { return localize('folderAlreadyAddedToProject', "A folder with the path '{0}' has already been added to the project", folderpath); }
 export function invalidInput(input: string) { return localize('invalidInput', "Invalid input: {0}", input); }
+export function invalidProjectPropertyValue(propertyName: string) { return localize('invalidPropertyValue', "Invalid value specified for the property '{0}' in .sqlproj file", propertyName); }
```
Comment (nahk-ivanov) [inline extensions/sql-database-projects/src/common/constants.ts:206]: Let me know if there is anything else that needs to be done for localization. I don't see these values in `package.nls.json`, so I didn't add this new one either.

File: extensions/sql-database-projects/src/models/project.ts [hunk 2]
```diff
@@ -508,18 +508,27 @@ export class Project implements ISqlProject {
 	}
 
 	public getProjectTargetVersion(): string {
-		// check for invalid DSP
-		if (this.projFileXmlDoc.getElementsByTagName(constants.DSP).length !== 1 || this.projFileXmlDoc.getElementsByTagName(constants.DSP)[0].childNodes.length !== 1) {
-			throw new Error(constants.invalidDataSchemaProvider);
+		let dsp: string | undefined;
+
+		try {
+			dsp = this.evaluateProjectPropertyValue(constants.DSP);
```
Comment (nahk-ivanov) [inline extensions/sql-database-projects/src/models/project.ts:514]: Ideally you would use `var` in here and get rid of the `let` line above completely, but linter in this project is configured to discourage `var`. :(

File: extensions/sql-database-projects/src/models/project.ts [hunk 3]
```diff
@@ -508,18 +508,27 @@ export class Project implements ISqlProject {
...
+		}
 
 		// get version from dsp, which is a string like Microsoft.Data.Tools.Schema.Sql.Sql130DatabaseSchemaProvider
-		// remove part before the number
-		let version: any = dsp.substring(constants.MicrosoftDatatoolsSchemaSqlSql.length);
-		// remove DatabaseSchemaProvider
-		version = version.substring(0, version.length - constants.databaseSchemaProvider.length);
+		// Remove prefix and suffix to only get the actual version number/name. For the example above the result
+		// should be just '130'.
+		const version =
```
Comment (nahk-ivanov) [inline extensions/sql-database-projects/src/models/project.ts:528]: I was not really sure what is going on here and why `substring` is called twice. I just made it start from where there first `substring` call was cutting it, instead of 0. Let me know, if I'm missing something.

File: extensions/sql-database-projects/src/models/project.ts [hunk 4]
```diff
@@ -529,6 +538,15 @@ export class Project implements ISqlProject {
 		return version;
 	}
 
+	/**
+	 * Gets the default database collation set in the project.
+	 *
+	 * @returns Default collation for the database set in the project.
+	 */
+	public getDatabaseDefaultCollation(): string {
+		return this.evaluateProjectPropertyValue(constants.DefaultCollationProperty, constants.DefaultCollation);
```
Comment (nahk-ivanov) [inline extensions/sql-database-projects/src/models/project.ts:547]: Here I'm not really sure what DacFX will do during deployment, if collation is not specified in the project - is it going to suggest `SQL_Latin1_General_CP1_CI_AS` or default to server-level (`master`) collation?

The UI in Visual Studio shows `SQL_Latin1_General_CP1_CI_AS` when project does not have the property and I decided that instead of having consumers to guess the default or hardcode General_CP1 everywhere - we will return it from here. This does imply that consumer will not be able to distinguish "property not set" vs. "property is set to SQL_Latin1_General_CP1_CI_AS", but same is true for VS users in the UI, so I assumed it's fair.

File: extensions/sql-database-projects/src/models/project.ts [hunk 5]
```diff
@@ -1031,6 +1049,52 @@ export class Project implements ISqlProject {
...
+	 */
+	private evaluateProjectPropertyValue(propertyName: string, defaultValue?: string): string | undefined {
+		// TODO: Currently we simply read the value of the first matching element. The code should be updated to:
+		//       1) Narrow it down to items under <PropertyGroup> only
+		//       2) Respect the `Condition` attribute on group and property itself
+		//       3) Evaluate any expressions within the property value
+
+		// Check if property is set in the project
+		const propertyElements = this.projFileXmlDoc.getElementsByTagName(propertyName);
+		if (propertyElements.length === 0) {
```
Comment (nahk-ivanov) [inline extensions/sql-database-projects/src/models/project.ts:1085]: There is a minor change of the behavior, compared to the logic that was hardcoded in the `getProjectTargetVersion`. Previously it was checking that there is only one element found, but now it just checks that there is at least something found and uses the first one. This makes it a little bit closer to reality, as there could be:
```xml
<PropertyGroup>
  <DatabaseCollation>foo</DatabaseCollation>
  <DatabaseCollation Condition="'$(UseCustomCollation)' == 'true'">bar</DatabaseCollation>
</PropertyGroup>
```
Which is a rather common pattern for properties initialization and we should support it, me thinks.

File: extensions/sql-database-projects/src/models/project.ts [hunk 6]
```diff
@@ -1031,6 +1049,52 @@ export class Project implements ISqlProject {
...
+			return defaultValue;
+		}
+
+		// Try to extract the value from the first matching element
+		const firstPropertyElement = propertyElements[0];
+		if (firstPropertyElement.childNodes.length !== 1) {
```
Comment (nahk-ivanov) [inline extensions/sql-database-projects/src/models/project.ts:1091]: This is the original logic and I didn't touch it, although I'm not sure if it's correct. This will still allow
```xml
<Prop1><foo/></Prop1>
```
which is probably not expected.

File: extensions/sql-database-projects/src/common/constants.ts (hunk 1 above)
Comment (Charles-Gagnon) [reply extensions/sql-database-projects/src/common/constants.ts:206]: There's two types of ways that things get localized - one is in package.nls.json. This is only for strings in a package.json. The other is through localize calls in the ts code. (In the end they both go through the same package though)

So what you have here is correct, and would be similar to any other localized strings that aren't in the package.json itself. 

(other resources outside of those two places are a whole separate matter and not something supported directly by ADS/VS Code)

File: extensions/sql-database-projects/src/models/project.ts (hunk 3 above)
Comment (kisantia) [reply extensions/sql-database-projects/src/models/project.ts:528]: it was just getting called twice for clarity on what was getting removed. Calling it once like this is fine

File: extensions/sql-database-projects/src/models/project.ts [hunk 7]
```diff
@@ -1031,6 +1049,52 @@ export class Project implements ISqlProject {
...
 		}
 	}
+
+	/**
+	 * Evaluates the value of the property item in the loaded project.
+	 *
+	 * @param propertyName Name of the property item to evaluate.
+	 * @returns Value of the property or `undefined`, if property is missing.
+	 */
+	private evaluateProjectPropertyValue(propertyName: string): string | undefined;
```
Comment (kisantia) [inline extensions/sql-database-projects/src/models/project.ts:1059]: why are these needed? Looks like defaultValue is already an optional parameter on line 1077

File: extensions/sql-database-projects/src/test/project.test.ts [hunk 8]
```diff
@@ -743,6 +745,55 @@ describe('Project: add SQLCMD Variables', function (): void {
...
+		const project = await Project.openProject(projFilePath);
+
+		should(project.getProjectTargetVersion()).equal('150');
+	});
+
+	it('Should throw on missing target database version', async function (): Promise<void> {
+		projFilePath = await testUtils.createTestSqlProjFile(baselines.sqlProjectMissingVersionBaseline);
+		const project = await Project.openProject(projFilePath);
+
+		should(() => project.getProjectTargetVersion()).throw("Invalid DSP in .sqlproj file");
```
Comment (kisantia) [inline extensions/sql-database-projects/src/test/project.test.ts:764]: use the error strings in constants.ts

**Review by kisantia — APPROVED** [review]

File: extensions/sql-database-projects/src/common/constants.ts [hunk 9]
```diff
@@ -277,6 +278,12 @@ export const ProjectGuid = 'ProjectGuid';
 export const Type = 'Type';
 export const ExternalStreamingJob: string = 'ExternalStreamingJob';
 
+/** Name of the property item in the project file that defines default database collation. */
```
[the remark this answers is not in the document]
Comment (nahk-ivanov) [reply extensions/sql-database-projects/src/common/constants.ts:281]: This is javadoc, not a single-line comment. :/

File: extensions/sql-database-projects/src/models/project.ts (hunk 7 above)
Comment (nahk-ivanov) [reply extensions/sql-database-projects/src/models/project.ts:1059]: This is how overloads are defined in TypeScript. The return type is different, depending on whether or not the default value provided. This eliminates the need of additional null check when default value is specified.

File: extensions/sql-database-projects/src/test/project.test.ts (hunk 8 above)
Comment (nahk-ivanov) [reply extensions/sql-database-projects/src/test/project.test.ts:764]: Hm, why? These are tests, not a production code. In tests we want to validate that we get the expected value. If the code is changing, then expected value should be updated. If you use a production constant in tests, then whenever the constant is updated to return an incorrect value - tests will not catch it.

If it helps to understand, then what you are suggesting is basically to write a test like `should(add(2, 2)).equal(add(2, 2))`. I don't think that is how tests are supposed to be written. Expected value simply cannot come from the production code that you are testing.

File: extensions/sql-database-projects/src/models/project.ts (hunk 2 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/models/project.ts:514]: Use of `var` in JS/TS has fallen out of favor, as `let` and `const` allow us to be more explicit with the scoping and mutability.

File: extensions/sql-database-projects/src/test/project.test.ts (hunk 8 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/test/project.test.ts:764]: I think there's a difference between the two depending on what the intention of the check is.  By using the error string, you're saying that the important thing is that the error gets thrown, not so much the specific string of the error; if the string gets updated, the test will - by design - still pass as long as the logic leading to throwing an error is intact.  By using a string literal, you're putting greater emphasis on the _contents_ of the string which, IMO, makes more sense when the string is processed in some way (as in your `add(2, 2)` example).  Don't think it matters too much one way or the other for this particular check.

**Review by Benjin — APPROVED** [review]

File: extensions/sql-database-projects/src/test/project.test.ts (hunk 8 above)
Comment (nahk-ivanov) [reply extensions/sql-database-projects/src/test/project.test.ts:764]: I think your explanation would make sense, if the ask was to remove the string altogether and only keep `.throw()` to validate that there is an error thrown, but the ask was to use production code in the expected value, which is a rather strange practice, to say the least.

File: extensions/sql-database-projects/src/test/project.test.ts [hunk 10]
```diff
@@ -225,7 +225,9 @@ describe('Project: sqlproj content operations', function (): void {
 		projFilePath = await testUtils.createTestSqlProjFile(baselines.newProjectFileBaseline);
 		const project = await Project.openProject(projFilePath);
 
-		await project.changeTargetPlatform('invalidPlatform');
```
Comment (nahk-ivanov) [inline extensions/sql-database-projects/src/test/project.test.ts:228]: @kisantia seems like you sneaked in the change before me... The question I have is whether any of the validation builds run these tests? I'd expect this test to fail after your change as `changeTargetPlatform` used to throw an exception in this case and as soon as you add `await` here - it will be an unexpected exception. Did you also change the logic in `changeTargetPlatform` not to throw (meaning test will now fail with my changed code) or how did the merge go through the validation?

Comment (nahk-ivanov) [reply extensions/sql-database-projects/src/test/project.test.ts:228]: Okay, so apparently `CI / linux` build runs extension tests and it is indeed now failing with my change because you changed the behavior and it doesn't throw an error any longer when set to incorrect platform. Was this intentional?
Looking at your changeset, this new behavior seems to come from the `getSystemDacpacUri`/`getSystemDacpacSsdtUri`, where you changed
```typescript
let version = this.getProjectTargetVersion();
```
to
```typescript
const versionFolder = this.getSystemDacpacFolderName();
```
and `getProjectTargetVersion` is what used to throw.

Can you please confirm what is the expected behavior for `changeTargetPlatform` when incorrect platform is provided?

File: extensions/sql-database-projects/src/test/project.test.ts (hunk 8 above)
Comment (Benjin) [reply extensions/sql-database-projects/src/test/project.test.ts:775]: Even then, when using the production constant, it still checks that the correct error is thrown (over a different error).  The only thing it doesn't protect from is someone accidentally checking in a mistake to that particular const (say, a typo or some copy pasta), but in doing so, it allows for single-line updates to the error message instead of requiring an edit in every test that confirms it.  Different, but not necessarily wrong or inferior, IMO.  Either way, feel free to leave as is since you obviously feel strongly about this.

Comment (nahk-ivanov) [reply extensions/sql-database-projects/src/test/project.test.ts:775]: > The only thing it doesn't protect from is someone accidentally checking in a mistake to that particular const (say, a typo or some copy pasta)

Unfortunately, that is not the only thing. In simple words, having the value that you expect in the test definition checks that the implementation in production code is correct. For instance, in this particular case, it also verifies that `localize` call performs as it should (or what matters for this particular codebase, as you don't want to test `localize` call - that strings are localized). The "constant" has an implementation and by using it here as the expected value you are basically claiming that production code does what it should and essentially should not be tested, which as I said is rather strange practice.

> it allows for single-line updates to the error message instead of requiring an edit in every test that confirms it

I think we should get on the same page here first. Writing and maintaining tests does imply an additional effort, sometimes this effort is times more, than the actual implementation. If additional effort incurred by tests is your concern than you should probably revisit the decision to have them at first place. Now, the solution to this particular problem is to introduce constant values in tests themselves, but never to reference production code as expected value.

> not necessarily wrong or inferior

I'd argue that it is in fact inferior, as it does not clearly explain the expectation. For instance, I'm currently having an issue with this PR, where expected behavior was never clearly set in the tests, and was updated in production code. This significantly diminishes the value of the tests at first place, as they can no longer serve as a reference point for the correct implementation. This is simply another example of that - if tomorrow somebody replaces the `localize` call with hard-coded string - your "different" implementation of the tests will pass, but in reality it should fail when tested for any non-"en-US" locale (which has 0% coverage today, by the way).

> feel free to leave as is since you obviously feel strongly about this.

Not only I plan to leave this as-is, but I'd also suggest your team to revisit basic coding practices.

File: extensions/sql-database-projects/src/test/project.test.ts (hunk 10 above)
Comment (kisantia) [reply extensions/sql-database-projects/src/test/project.test.ts:228]: I moved the Telemetry event in `changeTargetPlatform()` since it was sending the wrong from target version. Before the telemetry event was added, `changeTargetPlatform()` was not throwing when an invalid target platform was passed in and it was `getProjectTargetVersion()` that throws an error if the project version is invalid. This was done this way so that tests could easily set an invalid target platform and test the scenario of when a sqlproj has an invalid target platform.

Comment (nahk-ivanov) [reply extensions/sql-database-projects/src/test/project.test.ts:228]: Okay so what I'm hearing is that desired behavior is not to throw from `changeTargetPlatform` on invalid platforms, so I will update tests in my PR accordingly without going into another round of CS101, explaining why behavior of production code should not be driven by tests. 🤦
